lib/src/intl/intl_stream.dart

```diff
@@ -5,94 +5,52 @@
-
-  /// Return the next [howMany] items, or as many as there are remaining.
-  /// Does not modify the stream position.
-  dynamic peek([int howMany = 1]) {
-    dynamic result;
-    if (contents is String) {
-      String stringContents = contents;
-      result = stringContents.substring(
-          index, min(index + howMany, stringContents.length));
-    } else {
-      // Assume List
-      result = (contents as List).sublist(index, index + howMany);
-    }
-    return result;
-  }
+  /// Returns whether the input starts with [pattern] from the current index.
+  bool startsWith(String pattern) => contents.startsWith(pattern, index);
 
-  /// Return the remaining contents of the stream
-  dynamic rest() => peek(contents.length - index);
+  /// Return the next [howMany] characters, or as many as there are remaining,
+  /// without advancing the index.
+  String peek([int howMany = 1]) =>
+      contents.substring(index, min(index + howMany, contents.length));
 
-  /// Find the index of the first element for which [f] returns true.
-  /// Advances the stream to that position.
-  int? findIndex(bool Function(dynamic) f) {
-    while (!atEnd()) {
-      if (f(next())) return index - 1;
-    }
-    return null;
-  }
-
-  /// Find the indexes of all the elements for which [f] returns true.
-  /// Leaves the stream positioned at the end.
-  List<dynamic> findIndexes(bool Function(dynamic) f) {
-    var results = [];
-    while (!atEnd()) {
-      if (f(next())) results.add(index - 1);
-    }
-    return results;
-  }
+  /// Return the remaining contents of the String, without advancing the index.
+  String rest() => peek(contents.length - index);
 
-  /// Assuming that the contents are characters, read as many digits as we
-  /// can see and then return the corresponding integer, advancing the receiver.
+  /// Read as much content as [digitMatcher] matches from the current position,
+  /// and parse the result as an integer, advancing the index.
   ///
-  /// For non-ascii digits, the optional arguments are a regular expression
-  /// [digitMatcher] to find the next integer, and the codeUnit of the local
-  /// zero [zeroDigit].
-  int? nextInteger({RegExp? digitMatcher, int? zeroDigit}) {
-    var string = (digitMatcher ?? regexp.asciiDigitMatcher).stringMatch(rest());
+  /// The regular expression [digitMatcher] is used to find the substring which
+  /// matches an integer.
+  /// The codeUnit of the local zero [zeroDigit] is used to anchor the parsing
+  /// into digits.
+  int? nextInteger(RegExp digitMatcher, int zeroDigit) {
+    var string = digitMatcher.stringMatch(rest());
     if (string == null || string.isEmpty) return null;
     read(string.length);
-    if (zeroDigit != null && zeroDigit != constants.asciiZeroCodeUnit) {
+    if (zeroDigit != constants.asciiZeroCodeUnit) {
       // Trying to optimize this, as it might get called a lot.
       var oldDigits = string.codeUnits;
       var newDigits = List<int>.filled(string.length, 0);
```

Comment:
This filling confuses me - why not directly generate the `List` (or `Iterable`), for example
```
var newDigits = string.codeUnits.map((char) => char - zeroDigit + constants.asciiZeroCodeUnit);
```

Reply:
I know this is maybe out of scope for the review, I was just thinking as the name of the PR is "Simplify the IntlStream class"...

Reply:
I'd use an `Uint8List` to begin with. It's guaranteed to be ASCII digits, so it saves space, and the `String.fromCharCodes` is optimized for `Uint8List` input.

Reply:
The conversion below converts "other digits" to ASCII digits.

It assumes those other digits are base-10 (or lower, since they map to `0`..`9`), and that their code points are consecutive and in numerical order.
That might always be true, but it's something I'd like some evidence for.

Are non-ASCII digits really happening often enough that it's worth supporting?

Reply:
I'd like to refactor this to

```dart
      string = String.fromCharCodes(string.codeUnits
          .map((c) => c - zeroDigit + constants.asciiZeroCodeUnit));
```

And my initial naive benchmarks show it should be just as performant.

> I know this is maybe out of scope for the review, I was just thinking as the name of the PR is "Simplify the IntlStream class"...

Yeah, I'd break this one out into a separate change since the comment calls out perf as a motivation for the current implementation, and I don't know if we have benchmarks for it.

> Are non-ASCII digits really happening often enough that it's worth supporting?

I have no idea. Figuring out whether we can remove the support may be more effort than retaining it.